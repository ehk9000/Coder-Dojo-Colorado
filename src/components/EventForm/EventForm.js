import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import requests from '../../utils/requests/requests';

function EventForm({ 
  venueId, toggleView, event, updateAdminDash 
}) {
  const [eventDetails, setEventDetails] = useState({});
  const [startEndDate, setStartEndDate] = useState({});
  const [invalidField, setInvalidField] = useState('');

  const handleChange = ({ target }) => {
    const { name, value } = target;
    setEventDetails({ ...eventDetails, [name]: value });
  };

  const handleDateChange = (date, isEndTime) => {
    const startTime = new Date(date).toUTCString();
    const endTime = new Date(date).toUTCString();
    if (isEndTime) {
      setEventDetails({ ...eventDetails, endTime });
      setStartEndDate({ ...startEndDate, end: date });
    } else {
      setEventDetails({ ...eventDetails, startTime });
      setStartEndDate({ ...startEndDate, start: date });
    }
  };

  const preFillInputs = () => {
    if (event) {
      const { 
        startTime, endTime, name, notes, id
      } = event;
      const start = new Date(startTime);
      const end = new Date(endTime);
      setStartEndDate({ start, end });
      setEventDetails({ 
        name, notes, id, startTime, endTime
      });
    }
  };

  useEffect(() => {
    preFillInputs();
  }, []);

  const handleSubmit = async (e) => {
    e.preventDefault();
    setInvalidField('');
    if (!eventDetails.name || !eventDetails.startTime || !eventDetails.endTime) {
      setInvalidField('Please fill out required fields');
    } else {
      const newEvent = { ...eventDetails, venueId };
      let result;
      if (event) {
        result = await requests.editEvent(newEvent);
      } else result = await requests.createNewEvent(newEvent);
      if (result) {
        toggleView(false);
        updateAdminDash();
      }
    }
  };

  return (
    <div className="EventForm">
      <form onSubmit={handleSubmit}>
        <button type="button" className="cancel-event-btn" onClick={() => toggleView(false)}>&times;</button>
        {event ? <h2>Edit Event:</h2> : <h2>Create New Event:</h2>}
        <label htmlFor="event-name">
          Event Name:
          <input
            id="event-name"
            type="text"
            name="name"
            value={eventDetails.name || ''}
            onChange={handleChange}
          />
        </label>
        <DatePicker
          selected={startEndDate.start}
          onChange={date => handleDateChange(date, false)}
          minDate={new Date()}
          showTimeSelect
          dateFormat="MMMM d, yyyy h:mm aa"
          placeholderText="Starts"
          className="event-range"
        />
        <DatePicker
          selected={startEndDate.end}
          onChange={date => handleDateChange(date, true)}
          minDate={new Date()}
          showTimeSelect
          dateFormat="MMMM d, yyyy h:mm aa"
          placeholderText="Ends"
          className="event-range"
        />
        <label htmlFor="event-notes">
          Notes(Optional):
          <input
            id="event-notes"
            type="text"
            value={eventDetails.notes || ''}
            name="notes"
            onChange={handleChange}
          />
        </label>
        {invalidField && <p className="event-form-error">{invalidField}</p>}
        <button type="submit" className="create-event-btn" onClick={handleSubmit}>{event ? 'Update' : 'Create'}</button>
      </form>
    </div>
  );
}

EventForm.propTypes = {
  venueId: PropTypes.number,
  toggleView: PropTypes.func,
  updateAdminDash: PropTypes.func,
  event: PropTypes.object
};

EventForm.defaultProps = {
  venueId: 0,
  toggleView: () => {},
  updateAdminDash: () => {},
  event: {}
};


export default EventForm;
