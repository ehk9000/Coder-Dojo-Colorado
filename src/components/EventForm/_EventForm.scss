.EventForm {
  position: absolute;
  margin: auto;
  width: 80%;
  display: flex;
  justify-content: center;
  top: 20%;
  z-index: 2;

  form {
    width: 50%;
    background-color: red;
    padding: 20px;
    display: flex;
    flex-direction: column;
  }
}