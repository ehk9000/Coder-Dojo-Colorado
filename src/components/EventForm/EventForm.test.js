import React from 'react';
import { shallow } from 'enzyme';
import EventForm from './EventForm';

describe('EventForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<EventForm />);
  });

  it('should match snapshot', () => {
  });
});
