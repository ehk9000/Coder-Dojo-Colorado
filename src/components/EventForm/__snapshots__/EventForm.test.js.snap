// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm should match snapshot when mounted 1`] = `
<EventForm
  event={Object {}}
  toggleView={[MockFunction]}
  updateAdminDash={[MockFunction]}
  venueId={0}
>
  <div
    className="EventForm"
  >
    <form
      onSubmit={[Function]}
    >
      <button
        className="cancel-event-btn"
        onClick={[Function]}
        type="button"
      >
        ×
      </button>
      <h2>
        Edit Event:
      </h2>
      <label
        htmlFor="event-name"
      >
        Event Name:
        <input
          id="event-name"
          name="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <DatePicker
        allowSameDay={false}
        className="event-range start"
        dateFormat="MMMM d, yyyy h:mm aa"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        minDate={2019-08-24T06:00:00.000Z}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Starts"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-08-24T06:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={true}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      >
        <PopperComponent
          hidePopper={true}
          popperComponent={null}
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          popperPlacement="bottom-start"
          popperProps={Object {}}
          targetComponent={
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="event-range start"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Starts"
                readOnly={false}
                type="text"
                value="August 24, 2019 6:00 AM"
              />
            </div>
          }
        >
          <Manager>
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="event-range start"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Starts"
                      readOnly={false}
                      type="text"
                      value="August 24, 2019 6:00 AM"
                    />
                  </div>
                </div>
              </InnerReference>
            </Reference>
          </Manager>
        </PopperComponent>
      </DatePicker>
      <DatePicker
        allowSameDay={false}
        className="event-range end"
        dateFormat="MMMM d, yyyy h:mm aa"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        minDate={2019-08-24T06:00:00.000Z}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Ends"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-08-24T06:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={true}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      >
        <PopperComponent
          hidePopper={true}
          popperComponent={null}
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          popperPlacement="bottom-start"
          popperProps={Object {}}
          targetComponent={
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="event-range end"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Ends"
                readOnly={false}
                type="text"
                value="August 24, 2019 6:00 AM"
              />
            </div>
          }
        >
          <Manager>
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="event-range end"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Ends"
                      readOnly={false}
                      type="text"
                      value="August 24, 2019 6:00 AM"
                    />
                  </div>
                </div>
              </InnerReference>
            </Reference>
          </Manager>
        </PopperComponent>
      </DatePicker>
      <label
        htmlFor="event-notes"
      >
        Notes(Optional):
        <input
          id="event-notes"
          name="notes"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <button
        className="create-event-btn"
        onClick={[Function]}
        type="submit"
      >
        Update
      </button>
    </form>
  </div>
</EventForm>
`;

exports[`EventForm should match snapshot when mounted with event prop 1`] = `
<EventForm
  event={
    Object {
      "endTime": "2019-08-21 18:00:00 UTC",
      "eventCode": "45fbf7",
      "id": 2,
      "name": "Test Event 2",
      "notes": "testing notes",
      "startTime": "2019-08-21 17:00:00 UTC",
    }
  }
  toggleView={[Function]}
  updateAdminDash={[Function]}
  venueId={0}
>
  <div
    className="EventForm"
  >
    <form
      onSubmit={[Function]}
    >
      <button
        className="cancel-event-btn"
        onClick={[Function]}
        type="button"
      >
        ×
      </button>
      <h2>
        Edit Event:
      </h2>
      <label
        htmlFor="event-name"
      >
        Event Name:
        <input
          id="event-name"
          name="name"
          onChange={[Function]}
          type="text"
          value="Test Event 2"
        />
      </label>
      <DatePicker
        allowSameDay={false}
        className="event-range start"
        dateFormat="MMMM d, yyyy h:mm aa"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        minDate={2019-08-24T06:00:00.000Z}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Starts"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-08-24T06:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={true}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      >
        <PopperComponent
          hidePopper={true}
          popperComponent={null}
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          popperPlacement="bottom-start"
          popperProps={Object {}}
          targetComponent={
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="event-range start"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Starts"
                readOnly={false}
                type="text"
                value="August 24, 2019 6:00 AM"
              />
            </div>
          }
        >
          <Manager>
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="event-range start"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Starts"
                      readOnly={false}
                      type="text"
                      value="August 24, 2019 6:00 AM"
                    />
                  </div>
                </div>
              </InnerReference>
            </Reference>
          </Manager>
        </PopperComponent>
      </DatePicker>
      <DatePicker
        allowSameDay={false}
        className="event-range end"
        dateFormat="MMMM d, yyyy h:mm aa"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        minDate={2019-08-24T06:00:00.000Z}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Ends"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-08-24T06:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={true}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      >
        <PopperComponent
          hidePopper={true}
          popperComponent={null}
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          popperPlacement="bottom-start"
          popperProps={Object {}}
          targetComponent={
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="event-range end"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Ends"
                readOnly={false}
                type="text"
                value="August 24, 2019 6:00 AM"
              />
            </div>
          }
        >
          <Manager>
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="event-range end"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Ends"
                      readOnly={false}
                      type="text"
                      value="August 24, 2019 6:00 AM"
                    />
                  </div>
                </div>
              </InnerReference>
            </Reference>
          </Manager>
        </PopperComponent>
      </DatePicker>
      <label
        htmlFor="event-notes"
      >
        Notes(Optional):
        <input
          id="event-notes"
          name="notes"
          onChange={[Function]}
          type="text"
          value="testing notes"
        />
      </label>
      <button
        className="create-event-btn"
        onClick={[Function]}
        type="submit"
      >
        Update
      </button>
    </form>
  </div>
</EventForm>
`;
